### ISSUE TITLE
func ... publish failing with "Can not determine target framework for dotnet project"

### ISSUE THREAD

**Issue by im-samz** [Azure/azure-functions-core-tools#4209]
### Version

4.0.6610

### Description

`func start` seems to work fine, but `func azure functionapp publish <appName>` does not.

### Steps to reproduce

Repro:
1. Create a dotnet 8 function app in Azure
2. In VS C, create a HTTP trigger via a quick start template; use the default values
3. After the local project files are created, run `func start` --> this should work fine
4. Now run `func azure functionapp publish <appName>` --> this should fail

The deployment attempt will fail with: `"Can not determine target framework for dotnet project at $<PATH>"`

In --debug mode, I see the following error message:

```
Azure.Functions.Cli.Common.CliException: Can not determine target framework for dotnet project at $<PATH>
   at Azure.Functions.Cli.Helpers.DotnetHelpers.DetermineTargetFramework(String projectDirectory, String projectFilename) in D:\a\_work\1\s\src\Azure.Functions.Cli\Helpers\DotnetHelpers.cs:line 63
   at Azure.Functions.Cli.Actions.AzureActions.PublishFunctionAppAction.RunAsync() in D:\a\_work\1\s\src\Azure.Functions.Cli\Actions\AzureActions\PublishFunctionAppAction.cs:line 169
   at Azure.Functions.Cli.ConsoleApp.RunAsync[T](String[] args, IContainer container) in D:\a\_work\1\s\src\Azure.Functions.Cli\ConsoleApp.cs:line 63
```

**Comment by liliankasem** [comment]
Hi @im-samz, I was not able to repro your issue. I see that you used VS Code to create the Http Trigger function, but how did you create the function app itself? When you run this command, what result do you get?

`dotnet build <your-funcapp-project>.csproj -getproperty:TargetFramework`

For example, I get the `net8.0`

```log
dotnet build dni-http.csproj -getproperty:TargetFramework
net8.0
```


**Here are the steps I followed and was able to publish without issues:**

1. Create a dotnet 8 function app in Azure
2. In VS Code, create a new function app with a HTTP trigger function via a quick start template; use the default values
3. Run `func start` (everything runs as expected)
5. Run `func azure functionapp publish <appName>` (everything runs as expected)

```log
> func --version
4.0.6610

> func azure functionapp publish <myapp>

Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:04.75

Getting site publishing info...
[2024-12-09T18:28:15.934Z] Starting the function app deployment...
Creating archive for current directory...
Uploading 1.91 MB [###############################################################################]
Upload completed successfully.
Deployment completed successfully.
[2024-12-09T18:28:30.049Z] Syncing triggers...
Functions in func-dni8-publish-repro:
    HttpTrigger1 - [httpTrigger]
        Invoke url: https://<myapp>.azurewebsites.net/api/httptrigger1
```

**Comment by im-samz** [comment]
Hey @liliankasem, I created the function app through the Portal. I can confirm that it is `dotnet-isolated` running version `8.0`.

When I run `dotnet build <your-funcapp-project>.csproj -getproperty:TargetFramework`, I get the following:
```
Sams-MacBook-Pro:func-start szhang$ dotnet build func-start.csproj -getproperty:TargetFramework
MSBuild version 17.8.0-preview-23418-03+0125fc9fb for .NET
MSBUILD : error MSB1001: Unknown switch.
    Full command line: '/usr/local/share/dotnet/sdk/8.0.100-rc.1.23463.5/MSBuild.dll -maxcpucount -verbosity:m -restore -consoleloggerparameters:Summary func-start.csproj -getproperty:TargetFramework -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/usr/local/share/dotnet/sdk/8.0.100-rc.1.23463.5/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/usr/local/share/dotnet/sdk/8.0.100-rc.1.23463.5/dotnet.dll'
  Switches appended by response files:
Switch: -getproperty:TargetFramework

For switch syntax, type "MSBuild -help"
```
>Note, I am doing this on a Mac.

Now, if I go to the csproj file, I can confirm that TargetFramework is net8.0 `<TargetFramework>net8.0</TargetFramework>`

Interestingly, when I visit the Portal, I see this:
![Image](https://github.com/user-attachments/assets/cdf46123-e4a6-49de-9b35-b179caefc301)

I don't recall explicitly choosing "in-process". Could VSC have defaulted to "in-process", which is causing issues when running `func publish`?

**Comment by liliankasem** [comment]
> I created the function app through the Portal.

Sorry, I meant the function app project, not the azure resource. I.e. I used `func init` to create mine. Can you share a code repro (via a GH repository) of what you're trying to deploy?

> Note, I am doing this on a Mac.

Same

> Sams-MacBook-Pro:func-start szhang$ dotnet build func-start.csproj -getproperty:TargetFramework
MSBuild version 17.8.0-preview-23418-03+0125fc9fb for .NET
MSBUILD : error MSB1001: Unknown switch.
    Full command line: '/usr/local/share/dotnet/sdk/8.0.100-rc.1.23463.5/MSBuild.dll -maxcpucount -verbosity:m -restore -consoleloggerparameters:Summary func-start.csproj -getproperty:TargetFramework -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/usr/local/share/dotnet/sdk/8.0.100-rc.1.23463.5/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/usr/local/share/dotnet/sdk/8.0.100-rc.1.23463.5/dotnet.dll'
  Switches appended by response files:
Switch: -getproperty:TargetFramework

It looks like something is wrong with your dotnet installation?

`dotnet build <app>.csproj -getproperty:TargetFramework` is what core tools uses to determine the target framework so with this command not working on your device, core tools will not work to publish. Not sure that I can help with your .NET issue, might be worth reinstalling?

> MSBuild version 17.8.0-preview-23418-03+0125fc9fb for .NET

Maybe this version of MSBuild is the issue? I noticed mine is on version `17.12.7`

> Could VSC have defaulted to "in-process", which is causing issues when running func publish?

It should not default to in-proc anywhere, and I haven't observed it doing so, still if you're seeing that in the portal you may want to redeploy a new app and ensure isolated is selected. However, the issue/error you are seeing is actually happening before the deploy step, so I don't think this is connected.

**Comment by im-samz** [comment]
Hey Lilian! I created the project through the VSC command palete.

I updated my .NET SDK to the latest version, and now `dotnet build ...` is working! As a result, the `publish` command is working as well.

Seems like the issue was an outdated version. Sorry, I should've caught that earlier.

Thank you for helping! Glad it turned out to be an error on my end and not the platform.

**Comment by ColinAxians** [comment]
We got exactly the same error mentioned in the first post. Problem was we had a .net 7 function app and were using the newest version of azure functions core tools. The newest version of azure functions core tools uses command "dotnet build projectfilename.csproj -getproperty:TargetFramework" to get the framework version. This command does not work with .net 7 sdk. We downgraded azure functions core tools to version 4.0.5571 and publishing the function app worked again. (upgrade to newest .net version will of course be the next step).
